test/model/recent_senders_test.dart

```diff
@@ -0,0 +1,201 @@
+      ];
+
+      model.handleMessages(messages);
+      checkMatchesMessages(messages, model);
+    });
+
+    test('delete messages', () {
+      final model = RecentSenders();
+      final stream1 = eg.stream(streamId: 1);
+      final user1 = eg.user(userId: 1);
+      final user2 = eg.user(userId: 2);
+
+      final messages = [
+        eg.streamMessage(stream: stream1, topic: 'a', sender: user1, id: 300),
+        eg.streamMessage(stream: stream1, topic: 'b', sender: user1, id: 100),
+        eg.streamMessage(stream: stream1, topic: 'a', sender: user2, id: 200),
+      ];
+
+      model.handleMessages(messages);
+      checkMatchesMessages(messages, model);
+
+      model.handleDeleteMessageEvent(
+        DeleteMessageEvent(id: 1, messageIds: [300, 200],
+          messageType: MessageType.stream, streamId: stream1.streamId, topic: 'a'),
+        Map.fromEntries(messages.map((msg) => MapEntry(msg.id, msg))));
+
+      checkMatchesMessages(
+        [eg.streamMessage(stream: stream1, topic: 'b', sender: user1, id: 100)],
+        model);
+    });
+
+    test('latestMessageIdOfSenderInStream', () {
+      final model = RecentSenders();
+      final stream1 = eg.stream(streamId: 1);
+      final stream2 = eg.stream(streamId: 2);
+      final user1 = eg.user(userId: 1);
+      final user2 = eg.user(userId: 2);
+
+      final messages = [
+        eg.streamMessage(stream: stream1, sender: user1, id: 100),
+        eg.streamMessage(stream: stream1, sender: user1, id: 300),
+        eg.streamMessage(stream: stream1, sender: user1, id: 200),
+      ];
+
+      model.handleMessages(messages);
+
+      check(model)
+        ..latestMessageIdOfSenderInStream(
+            streamId: stream1.streamId, senderId: user1.userId).equals(300)
+        ..latestMessageIdOfSenderInStream(
+            streamId: stream1.streamId, senderId: user2.userId).equals(null)
+        ..latestMessageIdOfSenderInStream(
+            streamId: stream2.streamId, senderId: user1.userId).equals(null);
+    });
+
+    test('latestMessageIdOfSenderInTopic', () {
+      final model = RecentSenders();
+      final stream1 = eg.stream(streamId: 1);
+      final stream2 = eg.stream(streamId: 2);
+      final user1 = eg.user(userId: 1);
+      final user2 = eg.user(userId: 2);
+
+      final messages = [
+        eg.streamMessage(stream: stream1, topic: 'a', sender: user1, id: 300),
+        eg.streamMessage(stream: stream1, topic: 'a', sender: user1, id: 200),
+      ];
+
+      model.handleMessages(messages);
+
+      check(model)
+        ..latestMessageIdOfSenderInTopic(streamId: stream1.streamId,
+            topic: 'a', senderId: user1.userId).equals(300)
+        ..latestMessageIdOfSenderInTopic(streamId: stream1.streamId,
+            topic: 'a', senderId: user2.userId).equals(null)
+        ..latestMessageIdOfSenderInTopic(streamId: stream1.streamId,
+            topic: 'a', senderId: user2.userId).equals(null)
```

Comment:
This is the same check twice, right?

Let's take that as another indicator that these test cases (this and the `…InStream` case above) can be improved in their clarity. :slightly_smiling_face: When the test is really clear, it should be easy for you to scan what it's saying and hard to miss that two checks are actually the same.

Reply:
One other way to make these more compact and give them less visual repetition: since you're already specifying the stream IDs and user IDs explicitly above, just use the numbers here.

While you're at it, though: give the user IDs distinct values from the stream IDs. That'll avoid confusion for the reader, and also mean this would catch if the code under test somehow swapped the roles of stream ID and user ID. (That's why in my suggestion at https://github.com/zulip/zulip-flutter/pull/692#discussion_r1664653053 I had `streamId: 1` but `userId: 10`.)